src/NHibernate/Linq/Visitors/ParameterTypeLocator.cs

```diff
@@ -80,55 +101,68 @@ public static class ParameterTypeLocator
 			var visitor = new ConstantTypeLocatorVisitor(removeMappedAsCalls, targetType, parameters, sessionFactory);
 			queryModel.TransformExpressions(visitor.Visit);
 
-			foreach (var pair in visitor.ConstantExpressions)
+			var processedConstants = new HashSet<ConstantExpression>();
+			foreach (var pair in visitor.ParameterConstants)
 			{
-				var type = pair.Value;
-				var constantExpression = pair.Key;
-				if (!parameters.TryGetValue(constantExpression, out var namedParameter))
+				var namedParameter = pair.Key;
+				var constantExpressions = pair.Value;
+				// In case any of the constants has the type set, use it (e.g. MappedAs)
+				namedParameter.Type = constantExpressions.Select(o => visitor.ConstantExpressions[o]).FirstOrDefault(o => o != null);
+				if (namedParameter.Type != null)
 				{
 					continue;
 				}
 
-				if (type != null)
+				var parameterRelatedExpressions = new List<Expression>();
+				foreach (var expression in constantExpressions)
 				{
-					// MappedAs was used
-					namedParameter.Type = type;
-					continue;
+					if (visitor.RelatedExpressions.TryGetValue(expression, out var relatedExpressions))
+					{
+						parameterRelatedExpressions.AddRange(relatedExpressions);
+					}
 				}
 
+				var candidateTypes = new HashSet<IType>();
```

Comment:
Please extract `candidateTypes` calculation into a method.

Reply:
As far as I can see method can return either single `candidateType` or null in case multiple results/not found case. Also the whole routine can be skipped if `!FloatingPointNumericTypes.Contains(constantType)`